spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_structural.py

```diff
@@ -0,0 +1,514 @@
+
+
+class SynapseDynamicsStructural(AbstractSynapseDynamicsStructural):
+    def __init__(self, stdp_model=None, f_rew=10 ** 4, weight=0, delay=1,
+                 s_max=32,
+                 sigma_form_forward=2.5, sigma_form_lateral=1,
+                 p_form_forward=0.16, p_form_lateral=1,
+                 p_elim_dep=0.0245, p_elim_pot=1.36 * 10 ** -4,
+                 grid=np.array([16, 16]), lateral_inhibition=0,
+                 random_partner=False,
+                 seed=None):
+
+        AbstractSynapseDynamicsStructural.__init__(self)
+        self._f_rew = f_rew  # Hz
+        self._p_rew = 1. / self._f_rew  # ms
+        self._weight = weight
+        self._delay = delay
+        self._s_max = s_max  # maximum number of presynaptic neurons
+        self._lateral_inhibition = lateral_inhibition
+        self._sigma_form_forward = sigma_form_forward
+        self._sigma_form_lateral = sigma_form_lateral
+        self._p_form_forward = p_form_forward
+        self._p_form_lateral = p_form_lateral
+        self._p_elim_dep = p_elim_dep
+        self._p_elim_pot = p_elim_pot
+        self._grid = np.asarray(grid, dtype=int)
+        self._random_partner = random_partner
+        self._connections = {}
+
+        self.fudge_factor = 1.5
+        self._actual_row_max_length = self._s_max
+
+        if stdp_model is not None:
+            self.super = SynapseDynamicsSTDP(
+                timing_dependence=stdp_model.timing_dependence,
+                weight_dependence=stdp_model.weight_dependence,
+                dendritic_delay_fraction=stdp_model._dendritic_delay_fraction,
+                pad_to_length=self._s_max)
+        else:
+            self.super = SynapseDynamicsStatic(pad_to_length=self._s_max)
+
+        # Generate a seed for the RNG on chip that should be the same for all
+        # of the cores that have my learning rule
+        self._rng = np.random.RandomState(seed)
+        self._seeds = []
+        for _ in range(4):
+            self._seeds.append(self._rng.randint(0x7FFFFFFF))
+
+        # Addition information -- used for SDRAM usage
+        self._actual_sdram_usage = {}
+
+        self._ff_distance_probabilities = \
+            self.generate_distance_probability_array(
+                self._p_form_forward,
+                self._sigma_form_forward)
+        self._lat_distance_probabilities = \
+            self.generate_distance_probability_array(
+                self._p_form_lateral,
+                self._sigma_form_lateral)
+
+    @overrides(AbstractSynapseDynamicsStructural.get_parameter_names)
+    def get_parameter_names(self):
+        # TODO
+        return []
+
+    def distance(self, x0, x1, grid=np.asarray([16, 16]),
+                 type='euclidian'):  # @ReservedAssignment
+        x0 = np.asarray(x0)
+        x1 = np.asarray(x1)
+        delta = np.abs(x0 - x1)
+        #     delta = np.where(delta > grid * .5, delta - grid, delta)
+        #     print delta, grid
```

Comment:
This probably ought to be parenthesised, and `from __future__ import print_function` added to the (very!) top of the file. Except it's commented out…